### PR TITLE
move: report function names on abort

### PR DESCRIPTION
**Pull request by rvantonder**
See https://github.com/MystenLabs/sui/issues/7291 for description.

TL;DR abort errors print function names:

```patch
-Location: sui::dynamic_field in function definition 0 at offset 15, Abort Code: 0
+Location: sui::dynamic_field::add at offset 15, Abort Code: 0
```

This PR in Draft for discussion for now, just me reading code and coming up with a proposed fix.

What I found is that it's tricky to access function name info anywhere in Sui (without trying to get handles on all kinds of gnarly data structures), since error info like function names are only easily accessed in [move-lang/move](https://github.com/move-language/move/blob/0736cbc28a868c55c04fe892f46fe94d421eeb74/language/move-vm/runtime/src/interpreter.rs) and propagated from there. After some digging I found we're in a bit of luck because the error `message` is populated with function names for `MoveAbort`, and we can reuse those contents:

https://github.com/move-language/move/blob/0736cbc28a868c55c04fe892f46fe94d421eeb74/language/move-vm/runtime/src/interpreter.rs#L1985-L1995

This PR just uses those error message contents where possible and available, revealing the function names for `MoveAbort`. Main commit: [286382f](https://github.com/MystenLabs/sui/pull/7290/commits/286382f214e563aaceb033375280c92127127811)

We could also just parse the function name from the message, if we don't want to rely on the propagated error format.

A "proper" fix would probably mean exposing better typed values (e.g., function names) in [`VMError` in the `move-lang` repo](https://github.com/move-language/move/blob/0736cbc28a868c55c04fe892f46fe94d421eeb74/language/move-binary-format/src/errors.rs#L44-L56), but that looks like a super involved and painful change (and probably needs a bit more thought too). So error.message probably works OK.

There might be a way to do similar for `OUT_OF_GAS` errors, but I haven't gotten to that yet.

### CODE REVIEW

File: crates/sui-types/src/messages.rs [hunk 1]
```diff
@@ -1594,13 +1595,22 @@ impl Display for MoveLocation {
...
+                write!(f, "{s}")
+            }
+
+            None => {
+                write!(
+                    f,
+                    "{module} in function definition {function_idx} at offset {instruction}"
+                )
+            }
+        }
```
Comment (rvantonder) [inline crates/sui-types/src/messages.rs:1613]: There's gotta be a more terse way to do this but I'm still a Rust noob 😄

File: crates/sui-types/src/error.rs [hunk 2]
```diff
@@ -716,12 +716,13 @@ impl From<VMError> for ExecutionError {
...
                 debug_assert!(offset.is_some(), "Move should set the location on aborts");
-                let (function, instruction) = offset.unwrap_or((0, 0));
+                let (function_idx, instruction) = offset.unwrap_or((0, 0));
                 ExecutionFailureStatus::MoveAbort(
                     MoveLocation {
                         module: id.clone(),
-                        function,
+                        function_idx,
                         instruction,
+                        message: error.message().map(|m| m.into()),
```
Comment (rvantonder) [inline crates/sui-types/src/error.rs:725]: There's gotta be a more terse way to do this but I'm still a Rust noob 😄

File: crates/sui-types/src/messages.rs (hunk 1 above)
Comment (tnowacki) [reply crates/sui-types/src/messages.rs:1613]: This isn't bad :) It's a bit tricky with format strings. With a normal string you could use `Option`'s method `unwrap_or_else`. 

But the format strings aren't really values, so I don't think you can do that. 
You could do `if let Some(s) = message {} else {}` which might at least help reduce horizontal indents if that bothers you

File: crates/sui-types/src/messages.rs [hunk 3]
```diff
@@ -1594,13 +1595,22 @@ impl Display for MoveLocation {
...
         } = self;
-        write!(
-            f,
-            "{module} in function definition {function} at offset {instruction}"
-        )
+        match message {
+            Some(s) => {
+                write!(f, "{s}")
+            }
+
```
Comment (tnowacki) [inline crates/sui-types/src/messages.rs:1606]: I think it is uncommon to have a newline here, just since you mentioned being a Rust noob :)

File: crates/sui-types/src/messages.rs [hunk 4]
```diff
@@ -1363,8 +1363,9 @@ pub enum ExecutionFailureStatus {
 #[derive(Eq, PartialEq, Clone, Debug, Serialize, Deserialize, Hash)]
 pub struct MoveLocation {
     pub module: ModuleId,
-    pub function: u16,
+    pub function_idx: u16,
     pub instruction: CodeOffset,
+    pub message: Option<String>,
```
Comment (tnowacki) [inline crates/sui-types/src/messages.rs:1368]: Unfortunately, I don't think we want this string here. When we include a string in errors like this, it bakes in the string and all the formatting and everything into the protocol, which can make things a bit brittle. 

I'd be okay with including/resolving the string for the function name and including that in the `MoveLocation`. Thoughts?

**Review by tnowacki — CHANGES_REQUESTED** [review]

File: crates/sui-types/src/messages.rs (hunk 4 above)
Comment (rvantonder) [reply crates/sui-types/src/messages.rs:1368]: Makes sense! What comes to mind then is to parse out the function name from `error.message`? Still brittle but at least we control/gate/process the values we care about this way.

I don't see an obvious way to resolve the function name from other data structures in this area of the code (Sui side), unless you had a better proposal there? It seems really involved to propagate out via `VMError` in move-lang repo.

File: crates/sui-types/src/messages.rs (hunk 3 above)
Comment (rvantonder) [reply crates/sui-types/src/messages.rs:1606]: ah cool, looks like default `rustfmt` is happy either way with this newline being here or not. I'll prefer deleting it thanks :-)

**Comment by rvantonder** [comment]
closing in favor of https://github.com/MystenLabs/sui/pull/7306
